source/common/http/http2/codec_impl.cc

```diff
@@ -1004,6 +1007,10 @@ int ConnectionImpl::onInvalidFrame(int32_t stream_id, int error_code) {
 int ConnectionImpl::onBeforeFrameSend(const nghttp2_frame* frame) {
```

Comment:
This function is a callback for nghttp2 only for non-data frames. Given that you're accounting for all types of frame in the receive case, I'm guessing that wasn't the intent.

Trying instead `ConnectionImpl::onFrameSend`